pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -390,6 +398,25 @@ class Timestamp(_Timestamp):
             # User passed tzinfo instead of tz; avoid silently ignoring
             tz, tzinfo = tzinfo, None
 
+        # Allow fold only for unambiguous input
+        if (fold is not None and ts_input is not _no_input and
+                (is_integer_object(ts_input) or is_float_object(ts_input) or
+                    is_datetime64_object(ts_input) or isinstance(ts_input, str)
+                    or getattr(ts_input, 'tzinfo', None) is not None
+                )):
+            raise ValueError(
+                "Cannot pass fold with possibly unambiguous input: int, float, "
+                "numpy.datetime64, str, or timezone-aware datetime-like. "
+                "Pass naive datetime-like or build Timestamp from components."
+            )
+
+        if fold is not None and fold not in [0, 1]:
```

Comment:
can you integrate this with the above condition (e.g. they both hvae if fold is not None); that should be the fast path anyhow as that is the default

Reply:
Done.